src/osd/OSD.cc

```diff
@@ -7242,11 +7240,11 @@ void OSD::dispatch_context_transaction(PG::RecoveryCtx &ctx, PG *pg,
                                        ThreadPool::TPHandle *handle)
 {
   if (!ctx.transaction->empty()) {
-    ctx.on_applied->add(new ObjectStore::C_DeleteTransaction(ctx.transaction));
     int tr = store->queue_transaction(
       pg->osr.get(),
-      ctx.transaction, ctx.on_applied, ctx.on_safe, NULL,
+      std::move(*ctx.transaction), ctx.on_applied, ctx.on_safe, NULL,
       TrackedOpRef(), handle);
+    delete (ctx.transaction);
```

Comment:
Second thought, I am doing it for this and for ECBackend..Probably, it's good to change it now rather than later..Recovery io is also very important for our deployment (since each box is 512 TB) and the memory issues are more likely to hit in the recovery path..